airbyte-cdk/python/airbyte_cdk/destinations/vector_db_based/embedder.py

```diff
@@ -66,15 +69,25 @@ def check(self) -> Optional[str]:
             return format_exception(e)
         return None
 
-    def embed_chunks(self, chunks: List[Chunk]) -> List[List[float]]:
-        return self.embeddings.embed_documents([chunk.page_content for chunk in chunks])
+    def embed_chunks(self, chunks: List[Chunk], max_token_length: int) -> List[List[float]]:
+        embedding_batch_size = OPEN_AI_TOKEN_LIMIT // max_token_length
+        batches = create_chunks(chunks, batch_size=embedding_batch_size)
+        embeddings = []
+        for batch in batches:
+            embeddings.extend(self.embeddings.embed_documents([chunk.page_content for chunk in batch]))
+        return embeddings
```

Comment:
I'm struggling to understand this - a docstring could be helpful.

First, after a few reads, I came to these conclusions:

First, it looks like we are taking a conservative approach to number of chunks we can embed in each call - I think this is fine, but it's probably worth calling out in the docstring so behavior is clear. If max chunk length is 200 and token limit is 1,100, we'll send exactly 5 batches each time - even though some batches will end up smaller than the max token limit and we could have sent 6 or 7 (for instance).

Everywhere this class is inherited, it looks like we're overwriting `embed_chunks()` without calling the base class - and without performing the same limits. (`max_token_length` appears unused elsewhere.)

Reply:
Update: On reading through the entire file another time, it looks like the Azure and OpenAI subclasses actually do not override this method - the diff view was not making this clear.

I might suggest moving to a single implementation of `embed_chunks()`, where the behavior of not chunking could be driven by a zero or null in `max_token_length`.

Another approach would be to put `MAX_TOKEN_LIMIT` or `max_token_lenght` into the class as a class or object property - in which case the `embed_chunks()` probably doesn't need the `max_token_lenght` arg.

Reply:
One last comment here: I'm not seeing the rate limit around 'per second'. I see 'per batch'. Is it correct to infer that if the library will be _successful_ in automatically retrying batches, as long as they are lower than the match batch size per second. But if the batch has more tokens than the per-second limit allows, then we will retry infinitely and never succeed?

As I write it out, it makes more sense, but just want to confirm this is the correct understanding.

Reply:
> I might suggest moving to a single implementation of embed_chunks(), where the behavior of not chunking could be driven by a zero or null in max_token_length.

I'm not sure I follow here - there is just one embed_chunks for the openai-based embedders. The other embedder classes are doing a different thing, so they have different implementation. I don't think we should introduce a dependency to the `embeddings` instance as this is a langchain thing and I want to keep it on the level of the individual implementation as we might split it out anyway at some point.

> Another approach would be to put MAX_TOKEN_LIMIT or max_token_length into the class as a class constance, class property, or object property - in which case the embed_chunks() probably doesn't need to receive max_token_length arg - and again, the implementation could be identical.

We can add it as an object property, but it's dynamic for what the user configures, so it can't be a constant

> One last comment here: I'm not seeing the rate limit around 'per second'. I see 'per batch'. Is it correct to infer that if the library will be successful in automatically retrying batches, as long as they are lower than the match batch size per second. But if the batch has more tokens than the per-second limit allows, then we will retry infinitely and never succeed?

Yes, the regular retry is regulating if we are doing requests too fast. This fix is to make sure it works at all if the payload of the current batch of strings to embed is too large for the per-minute limit because the API will never complete such a request.